str_analysis/call_non_ref_pathogenic_motifs.py

```diff
@@ -0,0 +1,978 @@
+        locus_start_0based,
+        locus_end,
+        motif_size):
+    """Parses reads from the given bam/cram file that are within a window around the given STR locus. Then it computes
+    a list that, for each well-aligned read that overlaps the locus, contains the subsequence of bases from that read
+    that fall between the locus start and end coordinates. Also, it returns the total number of well-aligned read bases
+    to the left and to the right of the STR locus.
+
+    Args:
+        bam_or_cram_path (str): bam or cram path
+        reference_fasta (str): reference fasta path
+        locus_chrom (str): chromosome name of STR locus
+        locus_start_0based (int):  0-based start coordinate of STR locus
+        locus_end (int): end coordinate of STR locus
+        motif_size (int): motif size
+
+    Return:
+          overlapping_sequences, left_flank_n_well_aligned_bases, right_flank_n_well_aligned_bases
+    """
+    with pysam.Samfile(bam_or_cram_path, reference_filename=reference_fasta) as f:
+
+        # count reads in the left & right flanks to estimate read depth
+        # NOTE: f.fetch retrieves all reads that *overlap* the given interval
+        left_flank_n_well_aligned_bases = sum((r.query_alignment_length for r in f.fetch(
+            locus_chrom,
+            locus_start_0based - MARGIN - FLANK_SIZE,
+            locus_start_0based - MARGIN,
+            ) if r.mapq >= MIN_MAPQ))
+        right_flank_n_well_aligned_bases = sum((r.query_alignment_length for r in f.fetch(
+            locus_chrom,
+            locus_end + MARGIN,
+            locus_end + MARGIN + FLANK_SIZE,
+            ) if r.mapq >= MIN_MAPQ))
+
+        # get all sequences that overlap the locus (regardless of whether they're soft-clipped)
+        # pysam docs are @ https://pysam.readthedocs.io/en/latest/api.html#pysam.AlignedSegment.query_alignment_sequence
+        overlapping_sequences = []
+        for r in f.fetch(locus_chrom, locus_start_0based - MARGIN, locus_end + MARGIN):
+            if r.mapq < MIN_MAPQ:
+                # This filter discards MAPQ=0 reads. Although this may filter out some correctly-aligned informative
+                # reads, it also filters out reads that mis-mapped to this locus from other genomic regions that
+                # contain the reference motif. For example, for the RFC1 locus, positive control samples that are
+                # homozygous for the non-reference AAGGG motif may still have a few reads overlapping this
+                # locus that have the reference AAAAG motif and MAPQ=0. Not filtering out these reads would lead
+                # to this sample being incorrectly called heterozygous - having both the AAAAG and AAGGG motifs.
+                continue
+
+            read_sequence = r.seq
+            read_start_pos_including_soft_clips = r.reference_start - r.query_alignment_start
+            read_end_pos_including_soft_clips = read_start_pos_including_soft_clips + len(read_sequence)
```

Comment:
For me this made the code easier to understand because PySam behavior is complicated for these coordinates (for example, it includes soft-clips in some calculations and not in others). It might also be faster, but I'm not sure. 
Either way, this step shouldn't be a bottleneck.